### PR TITLE
Profiling

### PR DESCRIPTION
**Pull request by cmvcordova**
Basic notebook integrated with mkdocks-jupyter. Pushing to test changes on CI.

### CODE REVIEW

**Comment by lebrice** [comment]
Here is some feedback with respect to the notebook structure and the contents of the wandb report.
(the numbers in the list correspond to the same numbered points in this here: https://github.com/mila-iqia/ResearchTemplate/issues/11#issuecomment-2341723329)

1. Looks good!
2. First panel: Show metrics for a normal training run (with an rtx8000 GPU, ~1-4 CPUs), steps per second, cpu utilization, GPU utilization , RAM, VRAM. Could also be nice to display the # of CPUs, the CPU type, the GPU type, # of gpus, etc.
3. Second panel: Within the same job, show a comparison of the metrics above with / without training. (algorithm=no_op vs algorithm=example)
    3.1: Assuming that the dataloading is the bottleneck, make a plot that compares the throughput with different num_workers for the same number of CPUs (all run within the same job, the current interactive job.)
    3.2: Make a plot that compares the throughput of runs with the no-op algorithm with a different number of CPUs with either a fixed num_workers, or a fixed num_workers per cpu ratio.
    3.3: Given this better configuration for num_workers and n_cpus, show similar panels as in step 2, showing a comparison between A) previous parameters (lower bound), B) New, optimized parameters, and C) Optimized parameters without training (upper bound).
4. Comparing GPU vs CPU training:
    * The current panel looks good. The content will have to be updated to use the optimized # of cpus / num_workers from step 3.
    * Would be nice to add another panel showing the same comparison for training a small fcnetnetwork on MNIST. There, I suspect that the difference between GPU / CPU throughput shouldn't be that large. 
5. Comparing different types of GPUS
    * The panels look good! The content will have to be updated to use the optimized # of cpus / num_workers from step 3.
6. GPU Utilization
    * Also show the GPU utilization %, mem usage % in addition to the metrics in 2.

File: docs/examples/profiling.ipynb [hunk 1]
```diff
@@ -0,0 +1,622 @@
...
+   "metadata": {},
+   "source": [
+    "### Running a baseline and logging metrics on WandB"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In addition to specifying callbacks, the Mila Research template integrates using WandB as a logger, which enables the tracking of additional metrics through visualizations and dashboard creation. Given the flexibility and widespread adoption of the WandB logger, we'll be using it for the remainder of this tutorial, which will then be visualizable in the following [WandB report](https://wandb.ai/cesar-valdez-mcgill-university/ResearchTemplate/reports/Profiling--Vmlldzo5NDI1MjU0) , as supporting information for the experiments contained herein. We will now proceed to establish a baseline to profile, diagnose and optimize throughout this example."
```
Comment (lebrice) [inline docs/examples/profiling.ipynb:88]:
```suggestion
    "In addition to specifying callbacks, the Mila Research template integrates using WandB as a logger, which enables the tracking of additional metrics through visualizations and dashboard creation. Given the flexibility and widespread adoption of the WandB logger, we'll be using it for the remainder of this tutorial, which can be viewed in this [WandB report](https://wandb.ai/cesar-valdez-mcgill-university/ResearchTemplate/reports/Profiling--Vmlldzo5NDI1MjU0) , as supporting information for the experiments contained herein. We will now proceed to establish a baseline to profile, diagnose and optimize throughout this example."
```

File: docs/examples/profiling.ipynb [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%capture\n",
+    "!python project/main.py \\\n",
+    "    experiment=profiling \\\n",
+    "    trainer.logger.wandb.name=\"Baseline with training\" \\\n",
```
Comment (lebrice) [inline docs/examples/profiling.ipynb:100]:
```suggestion
    "    trainer.logger.wandb.name=\"Baseline\" \\\n",
```

File: docs/examples/profiling.ipynb [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Finding bottlenecks in your code is not necessarily clear or straightforward from the start. A sensible first step is to determine whether potential slowdowns originate from data loading or model computation. Running a model with and without training and contrasting the obtained outputs can help us determine whether the master process has a significant stall when fetching the next batch for training or not. Analyzing the difference between outputs can tell us the following about our model: \n",
+    "\n",
+    "- If the difference between data loading and training is close to 0, then the data loading procedure outpaces model computation, thus data loading is the bottleneck.\n",
+    "- If the difference between data loading and training is much greater than 0, then model computation outpaces data loading, thus computation is the bottleneck. \n",
+    "\n",
+    "We will proceed to run a series of experiments to identify potential bottlenecks: changing the workers involved in the dataloading process and the numbers of cpu assigned per task when training on a GPU."
```
Comment (lebrice) [inline docs/examples/profiling.ipynb:120]:
```suggestion
    "The first potential bottleneck to look out for is dataloading. An easy first step is to measure the throughput of your dataloading pipeline without any training. This can easily be done in this template with the `no_op` algorithm, which simply pulls batches from the dataloader, without doing any computation.  By Comparing the throughput (in samples/sec) that we get with the no-op algorithm vs our algorithm, we can infer the following: \n",
    "\n",
    "- If the throughput is much higher without training (e.g. >3x faster), then the slowest part of our code is the model computation. This is good.\n",
    "- If the difference in throughput (samples per second) between with and without training isn't significant, then data loading is the bottleneck, and we know where to focus our efforts to make the code more better. \n",
    "\n",
    "As can be seen in this figure (TODO: Add figure), we fall in the second case here: our dataloader is our bottleneck. Next, we will demonstrate how to improve the dataloading performance by changing the number of  workers and the numbers of cpus."
```

**Review by lebrice — CHANGES_REQUESTED** [review]

File: docs/install.md
Comment (lebrice) [inline docs/install.md:1]: remove

File: mkdocs.yml [hunk 4]
```diff
@@ -3,7 +3,9 @@ site_description: A project template and directory structure for Python data sci
 site_url: https://mila-iqia.github.io/ResearchTemplate/
 repo_url: https://www.github.com/mila-iqia/ResearchTemplate
 # edit_uri: edit/master/docs
-
+nav:
+  - Home: index.md
+  - Profiling your code: docs/examples/profiling.ipynb
```
Comment (lebrice) [inline mkdocs.yml:8]: Shouldn't this be added in the `index.md` file? I imagine that this would add the profiling notebook at the top level of the docs.

File: project/algorithms/example.py
Comment (lebrice) [inline project/algorithms/example.py:1]: remove

File: project/configs/algorithm/example.yaml
Comment (lebrice) [inline project/configs/algorithm/example.yaml:1]: remove

File: project/configs/algorithm/no_op.yaml
Comment (lebrice) [inline project/configs/algorithm/no_op.yaml:1]: This was probably already added in the `master` branch

File: project/configs/algorithm/optimizer/custom_adam.yaml
Comment (lebrice) [inline project/configs/algorithm/optimizer/custom_adam.yaml:1]: Was this already done on master?

File: project/configs/experiment/cluster_sweep_example.yaml [hunk 5]
```diff
@@ -1,4 +1,5 @@
 # @package _global_
+# yaml-language-server: $schema=../../../.schemas/experiment_cluster_sweep_example_schema.json
```
Comment (lebrice) [inline project/configs/experiment/cluster_sweep_example.yaml:2]:
```suggestion
```
This being here means that your vscode settings file isn't valid json (probably a trailing comma somewhere). This is the fallback for the auto-schema thing, in case it can't add the schema to the vscode settings.

File: project/configs/network/resnet18.yaml [hunk 6]
```diff
@@ -1,2 +1,2 @@
 _target_: torchvision.models.resnet18
-num_classes: ${instance_attr:datamodule.num_classes}
+num_classes: ${instance_attr:datamodule.num_classes,1000}
```
Comment (lebrice) [inline project/configs/network/resnet18.yaml:2]: Unrelated
```suggestion
num_classes: ${instance_attr:datamodule.num_classes}
```

File: project/configs/resources/one_gpu.yaml [hunk 7]
```diff
@@ -1,8 +1,8 @@
 # @package _global_
+# yaml-language-server: $schema=../../../.schemas/resources_one_gpu_schema.json
```
Comment (lebrice) [inline project/configs/resources/one_gpu.yaml:2]:
```suggestion
```

File: project/configs/trainer/default.yaml [hunk 8]
```diff
@@ -1,8 +1,9 @@
+# yaml-language-server: $schema=../../../.schemas/trainer_default_schema.json
```
Comment (lebrice) [inline project/configs/trainer/default.yaml:1]:
```suggestion
```

File: project/main.py [hunk 9]
```diff
@@ -53,23 +51,21 @@ def main(dict_config: DictConfig) -> dict:
         quiet=True,
         add_headers=False,
     )
-
     config: Config = resolve_dictconfig(dict_config)
-
+    # assert False, OmegaConf.resolve(dict_config)["datamodule"]
```
Comment (lebrice) [inline project/main.py:55]:
```suggestion
```

File: project/main.py [hunk 10]
```diff
@@ -88,8 +84,7 @@ def run(experiment: Experiment) -> tuple[str, float | None, dict]:
     )
 
     metric_name, error, metrics = evaluation(experiment)
-    if wandb.run:
-        wandb.finish()
```
Comment (lebrice) [inline project/main.py:92]: Make sure to add the `wandb.finish` somewhere though

File: project/main.py [hunk 11]
```diff
@@ -109,7 +104,7 @@ def evaluation(experiment: Experiment) -> tuple[str, float | None, dict]:
         # We want to report the training error.
         metrics = {
             **experiment.trainer.logged_metrics,
-            **experiment.trainer.callback_metrics,
+            **experiment.trainer._metrics,
```
Comment (lebrice) [inline project/main.py:107]: Why this change again? What's the difference between `_metrics` and `callback_metrics` ?

**Review by lebrice — CHANGES_REQUESTED** [review]

File: project/main.py (hunk 11 above)
Comment (cmvcordova) [reply project/main.py]: fixed in 2b138f6c550346ac96d62ed316c2aa6c820fa251

File: mkdocs.yml (hunk 4 above)
Comment (lebrice) [reply mkdocs.yml:8]: Edit: SUMMARY.md (which is the navigation bar), not `index.md`

Comment (cmvcordova) [reply mkdocs.yml]: fixed in 7d5c1d2e1c437ed76f58c76a78804bbc8c1cdb8c

File: project/main.py (hunk 10 above)
Comment (cmvcordova) [reply project/main.py:92]: I believe lines 60-68 contain the new wandb run logic - I may very well be missing something

File: docs/install.md
Comment (lebrice) [inline docs/install.md:1]: remove

File: project/algorithms/example.py
Comment (lebrice) [inline project/algorithms/example.py:1]: Isn't this already in master?

File: project/main.py [hunk 12]
```diff
@@ -39,7 +36,8 @@
 )
 def main(dict_config: DictConfig) -> dict:
     """Main entry point for training a model."""
-    print_config(dict_config, resolve=False)
+    # print_config(dict_config, resolve=False)
```
Comment (lebrice) [inline project/main.py:39]: I'd turn this off based on a config value instead of outright removing it
